### DISCUSSION TITLE
ESP32P4 support

### DISCUSSION THREAD

**Discussion by MrMarteng** [zephyrproject-rtos/zephyr#74958] [Q&A] [answered]
Is there a roadmap for integrating ESP32-P4 support into Zephyr?

**Comment by MrMarteng** [comment]
@sylvioalves Do you have some information on that topic?

**Reply by rftafas** [reply]
Hi,

No, it is not on immediate roadmap yet, but it will come. First there will be ESP-IDF 5.3 release (it's just a branch for now), then we have to update to hal_Espressif to 5.3. For now, we are still cleaning up after the changes prior to 3.7.

For the brave, there is a raw release for [esp-hal-3rdparty 5.2 with P4 preview](https://github.com/espressif/esp-hal-3rdparty/tree/sync/release_v5.2.a). The move from there is to put it on top of 5.1 and start working on Zephyr + P4. It is a lot of work, IMHO.

**Reply by LonerDan** [reply]
Hi,
out of curiosity, do you happen to know who/what drives development in the esp-hal-3rdparty repository and what are even the main changes in that code in reference to the corresponding esp-idf version?
From that repo's readme, it seems to me that it is mainly driven by efforts and needs by its consumer (they mention nuttx in there for example). 
The esp-idf's v5.3 is released since July of 2024, what exactly is needed for creation of `sync/release_v5.3` on [esp-hal-3rdparty](https://github.com/espressif/esp-hal-3rdparty)?

**Reply by sylvioalves** [reply]
Hi, there is an ongoing work to bring `sync/master.a` as `hal_espressif` basis. That will enable proper ESP32-P4 and other SoCs support on Zephyr.
As you have seen in the README file, some branches are automatically synchronized (with a custom script) with esp-idf release branches, such as `v5.1` and `master`. For the `master` branch we have `sync/master.a`.

**Reply by rftafas** [reply]
> out of curiosity, do you happen to know who/what drives development in the esp-hal-3rdparty repository and what are even the main changes in that code in reference to the corresponding esp-idf version?

Who: NuttX and Zephyr teams in Espressif.

What: necessity.
- Major bug fix on LL layer - which is very very unlikely, as this code was beaten up by the whole ESP-IDF team and users before we get it.
- It contains stuff we want. New chip, new chip revision, peripheral, new operating mode, etc.
- We jumped 5.2, 5.3 and 5.4 because there was no major additions to LL/HAL from the 5.1.4 and ESP32-P4 was not in mass production yet. The plan is only to support mass production devices.

> The esp-idf's v5.3 is released since July of 2024, what exactly is needed for creation of sync/release_v5.3 on [esp-hal-3rdparty](https://github.com/espressif/esp-hal-3rdparty)?
- What is needed: you can compare the _sync/release_5.1.c_ with the equivalent _hal_espressif_ (any _hal_espressif_ used by the 4.0~4.2 releases of Zephyr). The difference is what needs to be done.

**Comment by tannewt** [comment]
This has a roadmap from Espressif: https://developer.espressif.com/software/zephyr-support-status/

It says P4 is targeted for Zephyr 4.3 next month.

**Reply by rftafas** [reply]
Unfortunately, we are reviewing this schedule. It will be postponed to be available between v4.3 and v4.4 releases.

**Reply by tannewt** [reply]
It's still nice to have a central place to check!

**Reply by joesphan** [reply]
> Unfortunately, we are reviewing this schedule. It will be postponed to be available between v4.3 and v4.4 releases.

Hey just curious can I still count on this statement? Currently running against a limitation with freertos, next hw rev is using P4, want to migrate.

**Reply by zhanglazhuan** [reply]
@rftafas Hi, It's March now. Is the progress in line with expectations? Will v4.4 next month support P4 ?

**Reply by sylvioalves** [reply]
Bring up is completed, will push in a couple of days. It does it include support to all peripherals yet as we will gradually add it. Attention that feature freeze is in a few days and that could block having its support in the new release. I'll write back here when ready.

**Reply by zhanglazhuan** [reply]
It's really exciting to hear this news. Thank you for your excellent work

**Reply by DaAwesomeP** [reply]
Hey @sylvioalves, just checking in on the current timeline for the P4. Thanks!

**Reply by sylvioalves** [reply]
Hey, P4 is already working (basic peripherals only). Will submit its initial support during next week. However, merging is blocked until v4.4.0 release.

**Reply by zhanglazhuan** [reply]
> Hey, P4 is already working (basic peripherals only). Will submit its initial support during next week. However, merging is blocked until v4.4.0 release.

Hi, in the release doc (https://docs.zephyrproject.org/latest/releases/release-notes-4.4.html#boards-added-in-zephyr-4-4), I didn't see the P4 supported, and the boards/espressif has no esp32p4, so what is the situation now?

**Reply by zhanglazhuan** [reply]
> Hey, P4 is already working (basic peripherals only). Will submit its initial support during next week. However, merging is blocked until v4.4.0 release.

@sylvioalves Hi, any updates?

**Reply by rftafas** [reply]
Hi. It is under testing. Hal Espressif already supports it, however, it will have the similar set of peripherals as the other devices, which means that some of its differentiators won't be available in the first round.

**Reply by zhanglazhuan** [reply]
> Hi. It is under testing. Hal Espressif already supports it, however, it will have the similar set of peripherals as the other devices, which means that some of its differentiators won't be available in the first round.

@rftafas Hello, thanks for the update! 

I have a follow-up question regarding the statement: "Hal Espressif already supports it."
Currently, I cannot find any esp32p4 related files in my workspace, specifically in:
./soc/espressif
./modules/hal/espressif
./boards/espressif

Given that it's under testing, if I want to try out the basic peripherals early, what is the proper way to do it? Is there an active PR branch or an example repository I should pull from?

Ultimately, my biggest hope is to see the ESP32-P4X-Function-EV-Board added to ./boards/espressif in the near future. Thanks for the hard work!

**Reply by sylvioalves** [reply]
@zhanglazhuan exactly that:

```c
I (soc_init): ESP Simple boot
I (soc_init): compile time Apr 16 2026 23:47:53
W (soc_init): Unicore bootloader
I (soc_init): chip revision: v3.0
I (flash_init): SPI Speed      : 80MHz
I (flash_init): SPI Mode       : DIO
I (flash_init): SPI Flash Size : 16MB
I (boot): DRAM	: lma=00002020h vma=4ff20000h size=09884h ( 39044)
I (boot): DRAM	: lma=0000b8ach vma=4ff298a0h size=019c8h (  6600)
I (boot): IROM	: lma=00010000h vma=40800000h size=00d4ch (  3404)
I (boot): IROM	: lma=00020000h vma=40000000h size=05004h ( 20484)
I (boot): libc heap size 573 kB.
I (spi_flash): detected chip: gd
I (spi_flash): flash io: dio
*** Booting Zephyr OS build v4.4.0-453-gff50c7860e35 ***
Hello World! esp32p4_function_ev_board/esp32p4/hpcore
```

I am still cleaning the commits to submit it. Will be there soon.

**Reply by zhanglazhuan** [reply]
@sylvioalves Thanks for your work. My open-source hardware and system project is saved.

**Reply by razvanphp** [reply]
@sylvioalves looking forward to your release, where will you push it first? Thank you!

**Reply by qavaxb** [reply]
Is it possible to get some preview variant of port?

I would like to experiment with my P4 on Zephyr.

**Reply by imocanu** [reply]
Hi @sylvioalves, are your changes ready for latest Zephyr release ?

**Reply by pgeiem** [reply]
Hello @sylvioalves do you have any update on where you are and when you think this would be publicly available and usable ? Thanks for your works.

**Reply by rftafas** [reply]
PRs are being made. I believe much of the chip is usable. Did you try? It helps.

**Reply by pgeiem** [reply]
@rftafas no yet, will do for sure. Thanks

**Reply by sylvioalves** [reply, marked as answer]
Hey, it was merged. https://docs.zephyrproject.org/latest/boards/espressif/esp32p4_function_ev_board/doc/index.html

**Comment by dineshkrishna9999** [comment]
> Is there a roadmap for integrating ESP32-P4 support into Zephyr?

https://github.com/ksachdeva/zephyr-rtos-ai.git
